snuba/utils/streams/kafka.py

```diff
@@ -45,40 +82,99 @@ class KafkaConsumer(Consumer[TopicPartition, int, bytes]):
     )
 
     def __init__(self, configuration: Mapping[str, Any]) -> None:
-        self.__consumer = ConfluentConsumer(configuration)
-
-    def __wrap_assignment_callback(
-        self, callback: Callable[[Sequence[TopicPartition]], None]
-    ) -> Callable[[ConfluentConsumer, Sequence[ConfluentTopicPartition]], None]:
-        def wrapper(
-            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
-        ) -> None:
-            callback(
-                [TopicPartition(value.topic, value.partition) for value in partitions]
+        auto_offset_reset = configuration.get("auto.offset.reset", "largest")
+        if auto_offset_reset in {"smallest", "earliest", "beginning"}:
+            self.__resolve_partition_starting_offset = (
+                self.__resolve_partition_offset_earliest
+            )
+        elif auto_offset_reset in {"largest", "latest", "end"}:
+            self.__resolve_partition_starting_offset = (
+                self.__resolve_partition_offset_latest
+            )
+        elif auto_offset_reset == "error":
+            self.__resolve_partition_starting_offset = (
+                self.__resolve_partition_offset_error
             )
+        else:
+            raise ValueError("invalid value for 'auto.offset.reset' configuration")
+
+        # NOTE: Offsets are explicitly managed as part of the assignment
+        # callback, so preemptively resetting offsets is not enabled.
+        self.__consumer = ConfluentConsumer(
+            {**configuration, "auto.offset.reset": "error"}
+        )
+
+        self.__state = KafkaConsumerState.CONSUMING
 
-        return wrapper
+    def __resolve_partition_offset_earliest(
+        self, partition: ConfluentTopicPartition
+    ) -> ConfluentTopicPartition:
+        low, high = self.__consumer.get_watermark_offsets(partition)
+        return ConfluentTopicPartition(partition.topic, partition.partition, low)
+
+    def __resolve_partition_offset_latest(
+        self, partition: ConfluentTopicPartition
+    ) -> ConfluentTopicPartition:
+        low, high = self.__consumer.get_watermark_offsets(partition)
+        return ConfluentTopicPartition(partition.topic, partition.partition, high)
+
+    def __resolve_partition_offset_error(
+        self, partition: ConfluentTopicPartition
+    ) -> ConfluentTopicPartition:
+        raise ConsumerError("unable to resolve partition offsets")
 
     def subscribe(
         self,
         topics: Sequence[str],
         on_assign: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
         on_revoke: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
     ) -> None:
-        kwargs = {}
+        if self.__state is not KafkaConsumerState.CONSUMING:
+            raise InvalidState(self.__state)
```

Comment:
nit: since this logic happens in many methods, what about having a private method: 
```
__assertState(KafkaConsumerState.CONSUMING)
```

Just cosmetics, I think it makes it clear this is a precondition.

Reply:
I was on the fence about this, I'll add it.